File: java/src/main/java/com/threedimensionalloadingcvrp/validator/model/Item.java
package com.threedimensionalloadingcvrp.validator.model;

import lombok.Data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

/**
 * The type Item.
 */
@Data
public class Item {
    /** Unique Id */
    private int id;

    /** ItemType Id */
    private int type_id;

    /** Customer Id */
    private int customer_id;

    /** Rotation Flag (1: rotated (l = w, w = l)) */
    private boolean rotated;

    /** Length */
    private int l;

    /** Width */
    private int w;

    /** Height */
    private int h;

    /** Mass */
    private double mass;

    /** Fragility flag (1: fragile) */
    private boolean fragile;

    /** Maximum Load per Area (Load Bearing Strength) */
    private double lbs;

    /** Total Load */
    private double sumLoad;

    /** Position of Item, leftMost, deepest, bottommost corner */
    private Point min;

    /** Maximum Corner of Item */
    private Point max;

    // Calculated Members

    /** Percentage of direct Support of Items */
    private double support;

    /** Contains all ItemIds which are in Front */
    private HashSet<Integer> itemsFront;

    /** Contains all ItemIds which are Behind */
    private HashSet<Integer> itemsBehind;

    /** Contains all ItemIds which are Below */
    private HashSet<Integer> itemsBelow;

    /** Contains all ItemIds which are Above */
    private HashSet<Integer> itemsAbove;

    /** Current load per Area Unit */
    private List<Double> load;

    /**
     * Instantiates a new Item.
     *
     * @param id          the unique item id
     * @param customer_id the customer id
     * @param rotated     the rotation flag
     * @param type        the ItemType
     */
    public Item(final int id, final int customer_id, final boolean rotated, ItemType type) {
        this.id = id;
        this.customer_id = customer_id;
        this.rotated = rotated;
        this.type_id = type.getId();
        this.l = type.getL();
        this.w = type.getW();
        this.h = type.getH();
        this.mass = type.getMass();
        this.fragile = type.isFragile();
        this.lbs = type.getLbs();
        min = new Point(0, 0, 0);
        max = new Point(0, 0, 0);
        this.support = 0;
        this.sumLoad = 0;
        this.load = new ArrayList<>(Collections.nCopies(this.l * this.w, 0.0));
        itemsAbove = new HashSet<>();
        itemsBelow = new HashSet<>();
        itemsFront = new HashSet<>();
        itemsBehind = new HashSet<>();
    }

    // Functions
    /**
     * Sets the position of the Item.
     *
     * @param point the minimum corner point
     */
    public void setPosition(final Point point) {
        min = point;
        int maxX = !rotated ? min.getX() + l : min.getX() + w;
        int maxY = !rotated ? min.getY() + w : min.getY() + l;
        int maxZ = min.getZ() + h;
        max = new Point(maxX, maxY, maxZ);
    }

    /**
     * Is this item below another item.
     *
     * @param other the other item
     * @param touch is touching surface required
     * @return true if item is below another item.
     */
    public boolean isBelow(Item other, final boolean touch) {
        return (((touch && max.getZ() == other.min.getZ())
                || (!touch && max.getZ() <= other.min.getZ()))
		        && min.getY()       < other.max.getY()
		        && other.min.getY() < max.getY()
                && min.getX()       < other.max.getX()
		        && other.min.getX() < max.getX());
    }


    /**
     * Is this item above another item.
     *
     * @param other the other item
     * @param touch is touching surface required
     * @return true if item is above another item.
     */
    public boolean isAbove(Item other, final boolean touch) {
        return other.isBelow(this, touch);
    }

    /**
     * Is this item behind another item below.
     *
     * @param other the other item
     * @return true if item is behind another item.
     */
    public boolean isBehind(Item other) {
        return
                max.getX() <= other.min.getX() &&
                max.getZ() > other.min.getZ() &&
                min.getZ() < other.max.getZ() &&
                max.getY() > other.min.getY() &&
                min.getY() < other.max.getY();
    }

    /**
     * Is this item in front of another item.
     *
     * @param other the other item
     * @return true if item is in front of another item.
     */
    public boolean isInFront(Item other) {
        return other.isBehind(this);
    }
}

File: cpp/Validator/Item.cpp
#include "Item.h"

namespace validator {



	Item::Item(const int id, const int customer_id, const Rotation& rotation, const ItemType& type) {
		this->id = id;
		this->customer_id = customer_id;
		this->rotation = rotation;
		this->type_id = type.id;
		this->dimension = type.dimension;
		this->mass = type.mass;
		this->fragility = type.fragility;
		this->lbs = type.lbs;
		min = Point(0, 0, 0);
		max = Point(0, 0, 0);
		this->support = 0;
		this->sumLoad = 0.0;
		this->load = std::vector<double>(this->dimension.l * this->dimension.w, 0);
	}

	Item::Item() {
		min = max = Point(0, 0, 0);
		id = customer_id = mass = type_id = fragility = lbs = support = sumLoad = dimension.l = dimension.w = dimension.h = 0;
	}


	void Item::setPosition(Point point) {
		min = point;
		unsigned int maxX, maxY, maxZ;

		switch (rotation) {
		default:
			maxX = min.x + dimension.l;
			maxY = min.y + dimension.w;
			maxZ = min.z + dimension.h;
			break;

		case Rotation::LengthHeightWidth:
			maxX = min.x + dimension.l;
			maxY = min.y + dimension.h;
			maxZ = min.z + dimension.w;
			break;

		case Rotation::WidthLengthHeight:
			maxX = min.x + dimension.w;
			maxY = min.y + dimension.l;
			maxZ = min.z + dimension.h;
			break;

		case Rotation::WidthHeightLength:
			maxX = min.x + dimension.w;
			maxY = min.y + dimension.h;
			maxZ = min.z + dimension.l;
			break;


		case Rotation::HeightLengthWidth:
			maxX = min.x + dimension.h;
			maxY = min.y + dimension.l;
			maxZ = min.z + dimension.w;
			break;

		case Rotation::HeightWidthLength:
			maxX = min.x + dimension.h;
			maxY = min.y + dimension.w;
			maxZ = min.z + dimension.l;
			break;

		}

		max = Point(maxX, maxY, maxZ);
	}

	bool Item::isBelow(Item& other, const bool touch) const {
		if (this->id == other.id) return false;
		
		return (((touch && this->max.z == other.min.z) || (!touch && this->max.z <= other.min.z))
			&& this->min.y < other.max.y
			&& other.min.y < this->max.y
			&& this->min.x < other.max.x
			&& other.min.x < this->max.x);
	}

	bool Item::isAbove(Item& other, const bool touch) {
		return other.isBelow(*this, touch);
	}
	bool Item::isBehind(Item& other) {
		if (this->id == other.id) return false;

		return
			this->max.x <= other.min.x &&
			this->max.z > other.min.z &&
			this->min.z < other.max.z&&
			this->max.y > other.min.y &&
			this->min.y < other.max.y;
	}

	bool Item::isInFront(Item& other) {
		return other.isBehind(*this);
	}

}